File: backend/routes/history.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse

from models.scan import ScanResult
from store import history_store

router = APIRouter(prefix="/history", tags=["history"])


@router.get("")
async def list_history() -> list[dict]:
    return history_store.list_all()


@router.get("/{scan_id}", response_model=ScanResult)
async def get_history_scan(scan_id: str) -> ScanResult:
    result = history_store.get(scan_id)
    if not result:
        raise HTTPException(status_code=404, detail="Scan not found in history")
    return result


@router.get("/{scan_id}/export")
async def export_scan(scan_id: str) -> JSONResponse:
    result = history_store.get(scan_id)
    if not result:
        raise HTTPException(status_code=404, detail="Scan not found in history")

    export = {
        "scan_id": result.scan_id,
        "target": result.target,
        "command": result.command,
        "status": result.status.value,
        "started_at": result.started_at.isoformat(),
        "finished_at": result.finished_at.isoformat() if result.finished_at else None,
        "hosts": [h.model_dump() for h in result.hosts],
    }

    return JSONResponse(
        content=export,
        headers={
            "Content-Disposition": f'attachment; filename="nmap-vis-{result.target}-{result.scan_id[:8]}.json"'
        },
    )

File: frontend/src/api/history.ts
import { fetchApi } from './client'
import type { ScanResult } from '../types/scan'

export interface ScanSummary {
  scan_id: string
  target: string
  command: string
  status: string
  host_count: number
  port_count: number
  started_at: string
  finished_at: string | null
}

export function listHistory(): Promise<ScanSummary[]> {
  return fetchApi<ScanSummary[]>('/history')
}

export function getHistoryScan(scanId: string): Promise<ScanResult> {
  return fetchApi<ScanResult>(`/history/${scanId}`)
}
